editoast/src/views/timetable/train_schedule.rs

```diff
@@ -590,10 +608,344 @@ async fn occupancy_blocks(
+        &mut db_pool.get().await?,
+        valkey,
+        core_client,
+        &train_schedules,
+        &infra,
+        electrical_profile_set_id,
+    )
+    .await?;
+
+    let operational_point_track_offsets = operational_point.schema.track_offset();
+
+    let path_items = train_schedules
+        .iter()
+        .flat_map(|ts| ts.path.iter().map(|p| &p.location))
+        .collect::<Vec<_>>();
+
+    let path_item_cache =
+        PathItemCache::load(&mut db_pool.get().await?, infra_id, &path_items).await?;
+
+    let track_occupancy_map = {
+        let mut track_occupancy_map: HashMap<String, Vec<TrackOccupancy>> = HashMap::new();
+        for ((simulation, pathfinding), train_schedule) in
+            simulations_result.iter().zip(train_schedules)
+        {
+            // Get the track ranges from the pathfinding
+            let (track_ranges, path_item_positions) = match pathfinding {
+                PathfindingResult::Success(pathfinding_result_success) => (
+                    &pathfinding_result_success.track_section_ranges,
+                    &pathfinding_result_success.path_item_positions,
+                ),
+                PathfindingResult::Failure(_) => {
+                    // The timetable can have a pathfinding failure for a train schedule
+                    // because the train schedule is not valid.
+                    // We skip it because we don't want to compute the track occupancy for an invalid train schedule.
+                    // It's not a problem because we are just looking for the track occupancy of the operational point for the valid trains.
+                    tracing::info!(train_schedule.id, "pathfinding failed");
+                    continue;
+                }
+            };
+
+            let path_projection = PathProjection::new(track_ranges);
+
+            // Get the positions and the times from the simulation
+            let report_train = match simulation {
+                Response::Success { final_output, .. } => final_output.report_train.clone(),
+                _ => {
+                    tracing::info!(train_schedule.id, "simulation failed");
+                    continue;
+                }
+            };
+
+            let path_item_id = match_path_item_id_with_op_id(
+                &path_item_cache,
+                &train_schedule,
+                &operational_point_id,
+            );
+
+            if let Some((track_section, track_occupancy)) = compute_track_occupancy(
+                path_item_id,
+                path_item_positions,
+                &path_projection,
+                &operational_point_track_offsets,
+                &report_train,
+                &train_schedule,
+            ) {
+                track_occupancy_map
+                    .entry(track_section)
+                    .or_default()
+                    .push(track_occupancy)
+            }
+        }
+        track_occupancy_map
+    };
```

Comment:
This is a big block in an already big function. I've made https://github.com/OpenRailAssociation/osrd/pull/12098 as a proposition to extract a function out of it.